dhall/src/Dhall/DirectoryTree.hs

```diff
@@ -81,12 +125,63 @@ import qualified System.FilePath             as FilePath
     > ! "bar": null
     > ! "foo": "Hello"
 
-    This utility does not take care of type-checking and normalizing the
-    provided expression.  This will raise a `FilesystemError` exception upon
+    /Advanced construction of directory trees/
+
+    In addition to the ways described above using "simple" Dhall values to
+    construct the directory tree there is one based on a fixpoint encoding. It
+    works by passing a value of the following type to the interpreter:
+
+    > let User = < UserId : Natural | UserName : Text >
+    >
+    > let Group = < GroupId : Natural | GroupName : Text >
+    >
+    > let Access =
+    >       { execute : Optional Bool
+    >       , read : Optional Bool
+    >       , write : Optional Bool
+    >       }
+    >
+    > let Mode =
+    >       { user : Optional Access
+    >       , group : Optional Access
+    >       , other : Optional Access
+    >       }
+    >
+    > let Entry =
+    >       \(content : Type) ->
+    >         { name : Text
+    >         , content : content
+    >         , user : Optional User
+    >         , group : Optional Group
+    >         , mode : Optional Mode
+    >         }
+    >
+    > in  forall (r : Type) ->
+    >     forall  ( make
+    >             : { directory : Entry (List r) -> r
+    >               , file : Entry Text -> r
+    >               }
+    >             ) ->
+    >       List r
```

Comment:
Would it be possible to name the type variable `tree` both in the documentation and in the tests?

```suggestion
    > in  forall (tree : Type) ->
    >     forall  ( make
    >             : { directory : Entry (List tree) -> tree
    >               , file : Entry Text -> tree
    >               }
    >             ) ->
    >       List tree
```